src/textformats/cli/tf_decode.1.md
% TF\_DECODE(1) tf\_decode 1.0.0
% Giorgio Gonnella
% June 2021

# NAME

tf\_decode - decode text representation, according to given datatype definition

# SYNOPSIS

**tf_decode** string -s SPECFILE [-t DATATYPE] -e ENCODED\
**tf_decode** file [-s SPECFILE] [t DATATYPE] -i INFILE [-c S] [-w] [-x] [-u N]\

# DESCRIPTION

The input of the command is a string or a text file, which contains data in a
format described in a *TextFormats* specification.

The output of the command is the decoded data, in JSON format.

## Specifications

The specification file is either a YAML, JSON or compiled specification.

For the *file* subcommand, the specification may also be embedded, i.e.
provided, as YAML/JSON, in the beginning of the file, separated from the data
by a YAML document separator, i.e. a line consisting of --- only.

If a specification contains a datatype named 'default', this is used by default,
and the **\-\-datatype** option is not required.

## Scope section/file

If the **\-\-scope** option is set to 'section' or 'file', the definition must
be of kind 'composed\_of', 'list\_of' or 'named\_values', or a reference to
one of those. The separator must be the newline, and prefix and suffix must be
empty. Parsing is greedy, i.e. as many lines as possible are assigned to
each element of the compound datatype.

# OPTIONS

## Subcommands

**string**
: decode an encoded string and output as JSON

**file**
: decode a file using the specification

## Common options
**-s**, **\-\-specfile=**FILENAME
: specification file to use, YAML/JSON or compiled
  (not required for *file*, if the specification is embedded
  in the input file; always required for *string*)

**-t**, **\-\-datatype=**DATATYPE
: which datatype to use, among those defined by the specification
  (default: use datatype named 'default')

## Subcommand-specific options

*string* subcommand:

**-e**, **\-\-encoded=**STRING
: encoded string to be decoded
  (default: standard input, or empty string if none)

*file* subcommands:

**-i**, **\-\-infile=**FILENAME
: input file, containing the encoded data and, if **-s** is not used,
  an embedded specification (default: standard input; embedded specifications
  are not supported in this case)

**-w**, **\-\-wrapped**
: if set, and the datatype definition is 'one\_of', the decoded
  value is wrapped in a single-entry mapping, where the key
  is the name of the 'one\_of' branch which has been used for
  decoding, and the value is the unwrapped decoded value

**-x**, **\-\-splitted**
: if set, and scope is 'file' or 'section', the file/section
  structure is used for parsing but the decoded value of each
  line is output separately; advantage: it does not require to
  keep the whole file or file section in memory; scope 'line'
  is similar, but the structure of the file/section is not defined
  (default: false; ignored if scope is not 'file' or 'section')

**-c**, **\-\-scope=**S
: which part of the input file is targeted by the datatype
  definition; default: 'auto' (as specified by the scope key of
  the definition); other accepted values:
  'line' (each line of the file by itself),
  'unit' (units of fixed number of lines; see \-\-unitsize),
  'section' (sections of variable number of lines,
             as many as possible fit the definition, greedy),
  'file' (the entire file)

**-u**, **\-\-unitsize=**N
: how many lines does a unit contain (required if scope is 'unit'
  and no 'n\_lines' key is given in the definition; ignored if
  scope is not 'unit')

# EXIT VALUES
The exit code is 0 on success, anything else on error.



src/textformats/cli/tf_decode.nim
##
## Decode a string representation to the corresponding data,
## according to a given datatype from a specification
##
## The input is the string representation, either as the string itself
## or as a file in a given format
##
## The output is the decoded data, represented as JSON.
##

import tables, strutils, json, terminal
import ../../textformats
import cli_helpers

proc decode_string*(specfile: string, datatype = "default",
                    encoded = ""): int =
  ## decode an encoded string and output as JSON
  let to_decode = str_or_stdin(encoded)
  let definition = get_datatype_definition(specfile, datatype)
  try:
    echo $textformats.decode(to_decode, definition)
  except textformats.DecodingError:
    exit_with(ec_err_invalid_encoded, getCurrentExceptionMsg())

proc parse_scope_setting*(scope: string, dd: DatatypeDefinition): int =
  ##
  ## Compute the scope of a definition given a scope setting parameter (string)
  ##
  ## The scope setting parameter must be either one of the scope values
  ## (file, section, unit, line) or "auto". In the latter case, the
  ## scope must be defined in the datatype definition.
  ##
  if scope == "auto":
    if dd.get_scope == "undefined":
      exit_with(ec_err_setting,
         "Error: if no scope is provided (or is set to 'auto') a " &
         "'scope' key must be provided in the datatype definition")
  else:
    try:
      dd.set_scope(scope)
    except TextFormatsRuntimeError:
      exit_with(ec_err_setting,
                get_current_exception_msg() & "- auto\n")
  return 0

proc parse_unitsize_setting*(unitsize: int, dd: DatatypeDefinition): int =
  ##
  ## Compute the unit size of a definition given a setting parameter (int)
  ##
  ## If the unitsize setting parameter is 1, then the value is taken
  ## from the datatype definition (which must define it, in this case);
  ## otherwise the setting is used.
  ##
  if unitsize < 1:
    exit_with(ec_err_setting, "The 'unitsize' parameter must be >= 1")
  elif unitsize == 1 and dd.get_scope == "unit":
    if dd.get_unitsize > 1: return
    else:
      exit_with(ec_err_setting,
                "The 'unitsize' parameter for the scope 'unit' must be > 1")
  else:
    dd.set_unitsize(unitsize)
  return 0

proc decode_datafile*(specfile = "", datatype = "default", infile = "",
                      scope = "auto", splitted = false, wrapped = false,
                      unitsize = 1): int =
  ## decode a file, given a datatype definition
  let
    skip_embedded_spec = (specfile == "")
    specsrc = block:
      if skip_embedded_spec:
        if infile == "":
          exit_with(ec_err_setting,
                    "Please provide at least either the name of the " &
                    "specification file or of the input data file\n" &
                    "Specifications cannot be skip_embedded_spec in the " &
                    "standard input")
        fail_if_compiled(infile)
        infile
      else:
        specfile
    definition = get_datatype_definition(specsrc, datatype)
  var ec: int
  ec = parse_scope_setting(scope, definition)
  if (ec != 0): return ec
  ec = parse_unitsize_setting(unitsize, definition)
  if (ec != 0): return ec
  let level = if splitted: DplWhole else: DplLine
  try:
    if wrapped:
      definition.set_wrapped()
    decode_file(infile, definition, skip_embedded_spec,
                  decoded_processor_level = level)
  except textformats.DecodingError:
    exit_with(ec_err_invalid_encoded, getCurrentExceptionMsg())

when isMainModule:
  import cligen
  dispatch_multi([decode_string, cmdname = "string",
                  short = {"specfile": short_specfile,
                           "datatype": short_datatype,
                           "encoded":  short_encoded},
                  help = {"specfile": help_specfile,
                          "datatype": help_datatype,
                          "encoded":  help_encoded}],
                 [decode_datafile, cmdname = "file",
                  short = {"specfile": short_specfile,
                           "datatype": short_datatype,
                           "infile":  short_infile,
                           "scope": short_scope,
                           "splitted": short_splitted,
                           "wrapped": short_wrapped,
                           "unitsize": short_unitsize},
                  help = {"specfile": help_specfile_or_embedded,
                          "datatype": help_datatype,
                          "infile":  help_infile,
                          "scope": help_scope,
                          "splitted": help_splitted,
                          "wrapped": help_wrapped,
                          "unitsize": help_unitsize}])
